rohmu/transfer_pool.py

```diff
@@ -0,0 +1,220 @@
+
+
+_BASE_TRANSFER_INSTANCE_ATTRS = {"config_model", "log", "notifier", "prefix", "stats"}
+_BASE_TRANSFER_ATTRS = {attr for attr in vars(BaseTransfer) if not attr.startswith("__")} | _BASE_TRANSFER_INSTANCE_ATTRS
+
+# pylint: disable=abstract-method,super-init-not-called
+class SafeTransfer(BaseTransfer[StorageModel]):
+    """Helper class that helps the users in finding bugs in their code handling transfers.
+
+    This class prevents any call to a transfer instance that was returned to the pool.
+    It also logs a warning if the instance is garbage collected (returning it to the pool).
+
+    """
+
+    def __init__(self, item: TransferCacheItem, pool_reclaim: Callable[[TransferCacheItem], None]) -> None:
+        self._item = item
+        self._pool_reclaim = pool_reclaim
+        self._done = False
+
+    def __del__(self) -> None:
+        if not self._done:
+            LOG.warning("Transfer not marked as done was garbage collected. Returning it to the pool now.")
+            self.return_to_pool()
+
+    def __getattribute__(self, attr: str) -> Any:
+        if attr in _BASE_TRANSFER_ATTRS:
+            if self._done:
+                raise InvalidTransferError("Trying to access transfer instance already returned to the pool")
+            else:
+                return getattr(self._item.transfer, attr)
+        return super().__getattribute__(attr)
+
+    @classmethod
+    def from_model(cls, model: StorageModel) -> BaseTransfer[StorageModel]:
+        raise InvalidTransferError("You should not call class methods on SafeTransfer instances")
+
+    def return_to_pool(self) -> None:
+        """Return the underlying transfer instance to the pool.
+
+        After this call all uses of this transfer instance will raise InvalidTransferError.
+
+        This method can be called multiple times.
+        """
+        if self._done:
+            return
+        self._pool_reclaim(self._item)
+        self._done = True
+        del self._item
+
+
+class TransferPool:
+    def __init__(self, *, max_pool_age: float = TRANSFER_CACHE_MAX_AGE, max_transfer_age: float = TRANSFER_MAX_AGE):
+        self._max_pool_age = max_pool_age
+        self._max_transfer_age = max_transfer_age
+        self._caches: dict[str, _TransferCache] = {}
+        self._is_thread_safe: dict[str, bool] = {}
+        self._mutex = threading.Lock()
+
+    @contextmanager
+    def with_transfer(self, storage_config: dict[str, Any]) -> Generator[BaseTransfer[StorageModel], None, None]:
+        """Yield a transfer object according to the provided storage_config.
+
+        The transfer object may be created or cached. While the context manager is running the transfer object
+        won't be available to other threads.
+
+        Users of this contextmanager should NOT use the transfer object outside the `with` block
+        """
+        transfer = self._get(storage_config)
+        yield transfer.transfer
+        self._put(transfer)
```

Comment:
```suggestion
        transfer = self._get(storage_config)
        try:
            yield transfer.transfer
        finally:
            self._put(transfer)
```

Reply:
Fixed, also added a test checking that it works in case of exceptions